### PR TITLE
remove cacert.pem and use x509.SystemCertPool()

### PR DESCRIPTION
**Pull request by xulingqing**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
lgtm

File: pilot/pkg/model/jwks_resolver.go [hunk 1]
```diff
@@ -172,7 +172,7 @@ func newJwksResolverWithCABundlePaths(evictionDuration, refreshInterval time.Dur
 		},
 	}
 
-	caCertPool := x509.NewCertPool()
+	caCertPool := x509.SystemCertPool()
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:175]: the `SystemCertPool()` also returns an error variable: https://pkg.go.dev/crypto/x509#SystemCertPool. you can just log the error if any.

File: pilot/docker/Dockerfile.pilot [hunk 2]
```diff
@@ -15,7 +15,6 @@ FROM docker.io/istio/distroless:${BASE_VERSION} as distroless
 FROM ${BASE_DISTRIBUTION}
 
 COPY pilot-discovery /usr/local/bin/
-COPY cacert.pem /cacert.pem
```
Comment (yangminzhu) [inline pilot/docker/Dockerfile.pilot:18]: can you also update the Makefile so that the pilot docker target doesn't depend on the cacert.pem?

File: tools/istio-docker.mk [hunk 3]
```diff
@@ -96,10 +96,8 @@ docker.proxyv2: $(ISTIO_ENVOY_LINUX_RELEASE_DIR)/metadata-exchange-filter.wasm
 docker.proxyv2: $(ISTIO_ENVOY_LINUX_RELEASE_DIR)/metadata-exchange-filter.compiled.wasm
 	$(DOCKER_RULE)
 
-docker.pilot: BUILD_PRE=&& chmod 755 pilot-discovery cacert.pem
```
Comment (yangminzhu) [inline tools/istio-docker.mk:99]: just remove `cacert.pem`, not the whole line.

File: tests/integration/telemetry/outboundtrafficpolicy/helper.go [hunk 4]
```diff
@@ -361,7 +361,7 @@ func setupEcho(t *testing.T, ctx resource.Context, mode TrafficPolicy) (echo.Ins
 			},
 			TLSSettings: &common.TLSSettings{
 				// Echo has these test certs baked into the docker image
-				RootCert:   mustReadCert(t, "cacert.pem"),
+				RootCert:   "",
```
Comment (yangminzhu) [inline tests/integration/telemetry/outboundtrafficpolicy/helper.go:364]: 1. Just omit this field, the default value would be "".

2. Update the comment in TLSSettings.RootCert for the change that the system cert pool would be used if this is empty.

**Review by yangminzhu — CHANGES_REQUESTED** [review]

File: pkg/test/echo/client/client.go [hunk 5]
```diff
@@ -52,7 +52,13 @@ func New(address string, tlsSettings *common.TLSSettings) (*Instance, error) {
 			return nil, err
 		}
 
-		certPool := x509.NewCertPool()
+		var certPool *x509.CertPool
+		if tlsSettings.RootCert == "" {
+			certPool, _ = x509.SystemCertPool()
+		} else {
+			certPool.AppendCertsFromPEM([]byte(tlsSettings.RootCert))
```
Comment (yangminzhu) [inline pkg/test/echo/client/client.go:59]: the `certPool` is nil if RootCert != "", you should call NewCertPool() to create it first.

File: pkg/test/echo/common/model.go [hunk 6]
```diff
@@ -18,6 +18,7 @@ import "istio.io/istio/pkg/config/protocol"
 
 // TLSSettings defines TLS configuration for Echo server
 type TLSSettings struct {
+	// System cert pool would be used if RootCert field empty.
```
Comment (yangminzhu) [inline pkg/test/echo/common/model.go:21]: Change the comment to `RootCert supplies the root certificate to use inside the echo server, will use the default system cert pool if this is empty.`

**Review by myidpt — APPROVED** [review]

File: pkg/test/echo/client/client.go [hunk 7]
```diff
@@ -52,7 +52,13 @@ func New(address string, tlsSettings *common.TLSSettings) (*Instance, error) {
...
 		}
 
-		certPool := x509.NewCertPool()
+		var certPool *x509.CertPool
+		if tlsSettings.RootCert == "" {
+			certPool, _ = x509.SystemCertPool()
+		} else {
+			certPool = x509.NewCertPool()
+		}
+
```
[the remark this answers is not in the document]
Comment (yangminzhu) [reply pkg/test/echo/client/client.go:62]: You should choose between the 2 following, the current code is inconsistent with the comment:
1. you don't call AppendCertsFromPEM again in the line below as it's already called in line 60 or
2. you keep the following call, remove it from the above and change the `RootCert` comments to make it clear that the root cert is always appended to the system cert pool.

File: pkg/test/echo/common/model.go [hunk 8]
```diff
@@ -18,6 +18,7 @@ import "istio.io/istio/pkg/config/protocol"
 
 // TLSSettings defines TLS configuration for Echo server
 type TLSSettings struct {
+	//RootCert supplies the root certificate to use inside the echo server, will use the default system cert pool if this is empty.
```
Comment (yangminzhu) [inline pkg/test/echo/common/model.go:21]: Missed whitespace after `//`, and also see the other comment to see if this should be updated.

**Review by yangminzhu — CHANGES_REQUESTED** [review]

File: pkg/test/echo/common/model.go (hunk 8 above)
Comment (xulingqing) [reply pkg/test/echo/common/model.go]: I edited the code following the 2 option. Thanks a lot!

Comment (yangminzhu) [reply pkg/test/echo/common/model.go]: this looks better but still not what you're doing in the code. In your code you call NewCertPool if RootCert is empty, this means you're not always appending to the system cert pool.

If you want to make this always append, then just always call SystemCertPool() in the other place, there is no need to call NewCertPool() any more, and also change the comment to  `// If not empty, RootCert supplies the extra root cert that will be appended to the system cert pool.`

File: pkg/test/echo/client/client.go [hunk 9]
```diff
@@ -52,7 +52,11 @@ func New(address string, tlsSettings *common.TLSSettings) (*Instance, error) {
 			return nil, err
 		}
 
-		certPool := x509.NewCertPool()
+		var certPool *x509.CertPool
+		if tlsSettings.RootCert == "" {
+			certPool, _ = x509.SystemCertPool()
```
Comment (yangminzhu) [inline pkg/test/echo/client/client.go:57]: This still doesn't match your comment in model.go and has problems if RootCert != "" (certPool would be nil).

To implement as the comment you write in model.go, you should always call SystemCertPool() to get the certPool, and append tlsSettings.RootCert if it's not empty.

**Review by yangminzhu — CHANGES_REQUESTED** [review]

**Review by yangminzhu — APPROVED** [review]

**Review by bianpengyuan — APPROVED** [review]

File: pkg/test/echo/client/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xulingqing) [reply pkg/test/echo/client/client.go:60]: Thanks a lot for your instruction!

File: tests/integration/telemetry/outboundtrafficpolicy/helper.go [hunk 10]
```diff
@@ -374,7 +374,6 @@ func setupEcho(t *testing.T, ctx resource.Context, mode TrafficPolicy) (echo.Ins
 			},
 			TLSSettings: &common.TLSSettings{
 				// Echo has these test certs baked into the docker image
-				RootCert:   mustReadCert(t, "cacert.pem"),
```
Comment (howardjohn) [inline tests/integration/telemetry/outboundtrafficpolicy/helper.go:377]: Why is this removed? This is for the test image (dockerfile.app). I don't see why it should not have this (but keep the cert.crt and key?).

For pilot it clearly shouldn't, but the app image is meant exclusively for tests

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: tests/integration/telemetry/outboundtrafficpolicy/helper.go (hunk 10 above)
Comment (yangminzhu) [reply tests/integration/telemetry/outboundtrafficpolicy/helper.go:377]: it's replaced to also use the system cert pool just like the change in pilot, I don't know what is the reason for using cacert.pem even in test? It's just a forked version of the CA certs.

cert.crt is a self-signed cert and has nothing to do with cacert.pem but I'm not sure how it is used in the test.

Comment (howardjohn) [reply tests/integration/telemetry/outboundtrafficpolicy/helper.go:377]: I see. Well it seems clear the test is wrong then, maybe we use InsecureSkipVerify somewhere

**Review by howardjohn — APPROVED** [review]
